### PR TITLE
Fix bug with parsing a range expression.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/36122

Note: i believe this is a breaking change (if ranges have shipped).  That's because you *technically* could have written something like this before:

```c#
struct Whatever
{
     public static explicit operator Whatever(Range r) => ...;
}

//

(Whatever)..a
```

Previously, this would make a range, but then cast it to the Whatever type.  Now, this will properly think this is not a cast, changing the meaning to have the `(Whatever)` be the lower bound of the range.  I believe this is correct based on a reading of the spec (as well as beter matching a user's intuition):

Specifically:

> To resolve cast_expression ambiguities, the following rule exists: A sequence of one or more tokens (White space) enclosed in parentheses is considered the start of a cast_expression only if at least one of the following are true:
> 
> 1. ~not applicable~
> 2. The sequence of tokens is correct grammar for a type, and the token immediately following the closing parentheses is the token "~", the token "!", the token "(", an identifier (Unicode character escape sequences), a literal (Literals), or any keyword (Keywords) except as and is.

Since `..` is none of those tokens, this should not be a cast expression.

Now, if ranges are still in preview, then this is definitely safe to just fix right now.  If it's shipped, you'll have to make a breaking change call.  Personally, i think the risk is tiny so i think it's worth doing.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @gafter @jcouv @agocke @jaredpar @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -10385,6 +10385,7 @@ private static bool CanFollowCast(SyntaxKind kind)
                 case SyntaxKind.MinusMinusToken:
                 case SyntaxKind.OpenBracketToken:
                 case SyntaxKind.DotToken:
+                case SyntaxKind.DotDotToken:
                 case SyntaxKind.MinusGreaterThanToken:
                 case SyntaxKind.QuestionQuestionToken:
                 case SyntaxKind.EndOfFileToken:
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: note: i don't like this code here.  the spec is written clearly as a whitelist of tokens that cause us to think something is a cast expression.  However, the parser is written as a blacklist.  Meaning that any time we add new tokens (like dotdot) we need to know to fix things up here.  I'd really like to just make this more in line with the rules in https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#cast-expressions, but i'm worried about the risk of that.  If people want, i'd be willing to make another PR that tries that.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: @gafter for thoughts here.

**Comment by gafter** [comment]
I think https://github.com/dotnet/roslyn/issues/34483 should be fixed at the same time as https://github.com/dotnet/roslyn/issues/36122

**Comment by CyrusNajmabadi** [comment]
@gafter Do you want to take over this PR?

**Comment by gafter** [comment]
@CyrusNajmabadi According to https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md#c-80 @agocke is the engineer on this feature.

@agocke How do you want to proceed with this PR?

**Comment by CyrusNajmabadi** [comment]
Thanks @gafter !  Happy to let @agocke take over.  Thanks for making me aware of https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md#c-80

Andy, let me know what you'd like to do here.  We can take this separately if you prefer.  Or if you want to roll into any larger work you're doing, that's fine with me as well.  I don't feel comfortable taking all the precedence work that Neal mentioned simply because i don't think i have the time to dedicate to the appropriate amount of testing there.  

Thanks!

**Comment by agocke** [comment]
@CyrusNajmabadi I'll probably take your commit and work off of it in a separate PR, if you don't mind. That way I can massage things to the final shape, while still keeping your work.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I'll probably take your commit and work off of it in a separate PR, if you don't mind. That way I can massage things to the final shape, while still keeping your work.

Not a problem at all.  Feel free to close this out if you don't need it anymore.

**Comment by agocke** [comment]
I'm going to include this with an upcoming PR to fix precedence.

**Pull request closed without merging**